### PR TITLE
🔥 Feature: Add max size to cache

### PR DESCRIPTION
**Pull request by dranikpg**
This PR is a proposed solution to #1136 and allows **limiting the cache size**.

As stated in the issue, the cache should expire old entries to make room for new entries. To find old entries, it has to keep track of their expirations in a separate data structure. 

### Finding a solution

In general, the data structure has to support three main operations:
- adding entries 
- removing arbitrary entries (in case of self-expiration)
- removing the oldest entry (with lowest expiration point)

At first, it looks like a binary search tree would be the perfect fit. However GO's standart library doesn't offer much for implementing them and I didn't want to add new dependencies/piles of code. An easier and equally performant solution exists: lets use a **binary heap and track node movements for arbitrary removals**. This is what `indexedHeap` is for.

### Benchmarks

Max size does not affect read performance in any way

`Benchmark_Cache_MaxSize` is a benchmark for three cases:
- No max size specified (= no performance penalty)
- A very large max size (= only insertions)
- A small max size (= constant removal & insertion)

My results are as follows:
```
Benchmark_Cache_MaxSize/Disabled-16              1626 ns/op             349 B/op          7 allocs/op
Benchmark_Cache_MaxSize/Disabled-16              1588 ns/op             392 B/op          7 allocs/op
Benchmark_Cache_MaxSize/Unlim-16                 1812 ns/op             665 B/op          8 allocs/op
Benchmark_Cache_MaxSize/Unlim-16                 1678 ns/op             629 B/op          8 allocs/op
Benchmark_Cache_MaxSize/LowBounded-16            1178 ns/op             256 B/op          8 allocs/op
Benchmark_Cache_MaxSize/LowBounded-16            1190 ns/op             256 B/op          8 allocs/op
```

We're **up one allocation from first to second case and about 10% slower** per insertion. The third case is the fastest because our cache always stays small.

This is a benchmark where the handler takes zero time to execute. 10% might sound like much, but the **difference is less than one microsecond in my case** (thats actually suspiciously small 🤔). I doubt any handler call that has to be cached would care about such a time period.

### Measuring real memory consumption

Instead of counting entries, we could count their body size to measure real memory consumption. 

### Other possible soltuions

* We could ignore expiration all together and remove just the oldest entry by time of insertion. This would be the easiest and most performant solution. I'm not sure whether this would be correct behaviour (what if we pre-cache a bunch of big pages on startup - should we drop them all?)
* Implement some garbage collection goroutine. This sounds more complex and is full of bad worst cases

### CODE REVIEW

**Comment by ReneWerner87** [comment]
Thanks for this feature

will review it this weekend and if everything is ok will merge it

File: middleware/cache/config.go [hunk 1]
```diff
@@ -59,6 +59,13 @@ type Config struct {
 	//
 	// Default: false
 	StoreResponseHeaders bool
+
+	// Max number of entries simultaneously stored in cache. When limit is reached,
+	// entries with the nearest expiration are deleted to make room for new.
+	// 0 means no limit
+	//
+	// Default: 0
+	MaxSize int
```
Comment (ReneWerner87) [inline middleware/cache/config.go:68]: can you use uint -> negative sizes are not possible

File: middleware/cache/cache.go [hunk 2]
```diff
@@ -59,6 +59,8 @@ func New(config ...Config) fiber.Handler {
 	)
 	// Create manager to simplify storage operations ( see manager.go )
 	manager := newManager(cfg.Storage)
+	// Create indexed heap for tracking expirations ( see heap.go )
+	heap := &indexedHeap{}
```
Comment (ReneWerner87) [inline middleware/cache/cache.go:63]: the allocations come from the fact that elements must be created here at runtime.

could you not create the array with fixed size already in the heapmap with the information about the max size ?

**Comment by dranikpg** [comment]
I've just tracked down that one allocation 🥳 
It comes from promoting a `heapEntry` to an `interface{}` when pushing onto the heap.
Pushing manually (append + heap.Fix) avoids it.

**Comment by ReneWerner87** [comment]
I just remembered that we should use mutex, sync map or atomic due to the concurrency processes to prevent race conditions in the access

File: middleware/cache/cache.go (hunk 2 above)
Comment (ReneWerner87) [reply middleware/cache/cache.go:63]: Yes you are right

**Comment by ReneWerner87** [comment]
Can you update the documentation for the middleware in our docs repository

**Comment by dranikpg** [comment]
> I just remembered that we should use mutex, sync map or atomic due to the concurrency processes to prevent race conditions in the access

Sync is already in place for storage writes, heap modifications happen around it. There shouldn't be any issues

**Comment by dranikpg** [comment]
> * [ ]  Pls update the documentation for the middleware in our docs repository
> * [ ]  update the config documentation in our README.md for the middleware

Before touching docs: are we sure we'll count entries and not bytes?

**Comment by ReneWerner87** [comment]
> > I just remembered that we should use mutex, sync map or atomic due to the concurrency processes to prevent race conditions in the access
> 
> Sync is already in place for storage writes, heap modifications happen around it. There shouldn't be any issues
> 

Perfect

**Comment by ReneWerner87** [comment]
> > * [ ]  Pls update the documentation for the middleware in our docs repository
> > * [ ]  update the config documentation in our README.md for the middleware
> 
> Before touching docs: are we sure we'll count entries and not bytes? 

Will look tomorrow again more closely and answer

**Comment by ReneWerner87** [comment]
@dranikpg 
Ok then we should name the configuration flag differently

e.g. maxEntries

@efectn @hi019 what do you think about the count of bytes vs count of entries

**Comment by efectn** [comment]
> @dranikpg Ok then we should name the configuration flag differently
> 
> e.g. maxEntries
> 
> @efectn @hi019 what do you think about the count of bytes vs count of entries

I think counting bytes is better option

File: middleware/cache/heap.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
+package cache
+
+import (
+	"container/heap"
+)
+
+type heapEntry struct {
+	key   string
+	exp   uint64
+	bytes uint
```
Comment (dranikpg) [inline middleware/cache/heap.go:10]: Entry size is stored in the heap to update the total size without reading entries on delete

**Review by efectn — APPROVED** [review]

**Comment by ReneWerner87** [comment]
@dranikpg thx
https://github.com/gofiber/fiber/pull/1892#issuecomment-1117638432
don´t forget the doc repository

**Review by ReneWerner87 — APPROVED** [review]

**Pull request merged**